Type Inference

LensPlaysGames opened the pull request:
Here's the thought process.

```lisp
a : integer = 69
```

Because the compiler already knows the type of `69` (`t_integer_literal`), there is not really any reason the programmer should have to explicitly state the type of the variable (unless a conversion is wanted, in which case type casting should be used anyway). This means we can effectively reduce the length of most declarations, given an initalizer.

```lisp
a :: 69
```

This code has the same meaning as above, but with the added benefit of being shorter, and not any less understandable. `a` is a variable declared as the type of `69` (which everyone knows is an integer).


Luckily, this isn't *too* complicated to implement. I'm sure I've made subtle mistakes, but overall it seems to be working well.

LensPlaysGames committed:
¡Type Inference!

Well, that was easy...

LensPlaysGames committed:
[Sema] Turns out preserving integer_literals is important

We need to have a pass right in between sema and codegen that promotes
integer literals or something

LensPlaysGames committed:
[Sema] Indentation

Sirraide commented on src/typechecker.c:
```diff
@@ -813,7 +813,14 @@ NODISCARD bool typecheck_expression(AST *ast, Node *expr) {
+        // Type inference :^)
+        if (!expr->type) {
+          expr->type = expr->declaration.init->type;
+          // TODO: Maybe this shouldn't happen here? A pass after the
+          // bulk of sema that reduces all leftover integer_literals to
+          // integers may be a better choice.
+          if (expr->type == t_integer_literal) expr->type = t_integer;
```
I mean, if you’re *inferring* the type, is there ever a situation where preserving the type as `integer_literal` would be of any use?

LensPlaysGames replied:
Probably not, to be honest. I was just thinking that maybe this decl would be better suited as a `byte` or some other convertible integer type based on a later use. You're right that I can probably just remove this todo

LensPlaysGames committed:
[Tests] Type inference tests for `integer` and `@byte`

Also some comments shifted around and there was some removal of
redundant code.

LensPlaysGames committed:
[Minor/Sema] Typo
